content/blog/next-level-iac-briding-the-declarative-gap/index.md

```diff
@@ -0,0 +1,164 @@
+---
+title: "Next-level IaC: Bridging the Declarative Gap"
+
+date: 2024-06-07
+draft: false
+
+meta_desc: "Pulumi bridges the worlds of declarative and standard programming using asynchronous semantics."
+meta_image: meta.png
+authors:
+    - troy-howard
+tags:
+    - next-level-iac
+    - typescript
+    - python
+    - architecture
+---
+
+Pulumi stands out in the world of infrastructure-as-code (IaC) for its flexibility and ease of use. This is due to its unique approach of using general purpose programming languages to describe system configuration, rather than flat configuration files or domain-specific languages (DSL). It accomplishes this through a small but brilliant implementation detail that bridges the worlds of declarative and standard programming using asynchronous semantics.
+
+<!--more-->
+
+In Pulumi, you describe your infrastructure in code – real code, not a [DSL][dsl-wiki] or [YAML][yaml-wiki] – using your preferred general purpose programming language You don't have to become a specialist in a niche proprietary declarative language, like Terraform's HCL. Rather, you can reuse your existing programming skills, writing in standard imperative, object-oriented, and even functional language styles, while still gaining all of the benefits of the declarative style that other tools emphasize.
+
+Pulumi provides a unique mix of a [declarative model][pulumi-declarative-imperative-docs] embedded and implemented inside of a standard programming language, allowing all the flexibility of custom imperative code, while still enabling Pulumi's [deployment engine][pulumi-engine-docs] to infer opportunities for parallel asynchronous execution and to converge a partially-realized system.
+
+The secret sauce that lets Pulumi achieve this is the [`apply`][apply-docs] function, and the way it handles inputs and outputs via asynchronous programming constructs, similar to [futures][futures-wiki]. We'll go deeper on that in this article, but first let's talk a little bit about the flow of operations in Pulumi.
```

Comment:
might want to say "future or promises" since promises might be the more well known term

Reply:
good call. i updated it...  the semantics are fuzzy, but IMO Pulumi Inputs/Outputs are technically futures, not promises, but these terms are so often confused/interchanged that in our TypeScript implementation, "Promises" are used under the hood: https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/output.ts#L914 ... and https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/output.ts#L778 are demonstrative. So, really the distinction between the concepts, even in our own code, is unclear.